src/zeek.bif

```diff
@@ -922,6 +922,41 @@ function paraglob_equals%(p_one: opaque of paraglob, p_two: opaque of paraglob%)
 	);
 	%}
 
+%%{
+enum FNV_HASH_TYPE { FNV1a32, FNV1a64};
+
+uint64_t fnva(zeek::Val* input, uint64_t offset, uint64_t prime, FNV_HASH_TYPE type ) {
+	zeek::ODesc desc(zeek::DESC_READABLE);
+
+	auto length = 0;
+	const u_char* bytes = NULL;
+
+	if (input->GetType()->Tag() == zeek::TYPE_STRING)
+		{
+			length = ((zeek::StringVal*) input) -> Len();
+			bytes = ((zeek::StringVal*) input)-> Bytes();
+		}
+	else
+		{
+			input->Describe(&desc);
+			bytes = desc.Bytes();
+			length = desc.Len();
+		}
+
+	uint64_t rval = offset;
+	for ( auto i = 0; i < length; ++i )
+	{
+		rval ^= (uint64_t) bytes[i];
+		rval *= prime;
+
+		if (type == FNV1a32)
+			rval = rval % (1ULL << 32);
```

Comment:
You could template fnva with the type and skip this I think - also allows to drop the type argument.

```
-uint64_t fnva(zeek::Val* input, uint64_t offset, uint64_t prime, FNV_HASH_TYPE type ) {
+template<typename IntType>
+IntType fnva(zeek::Val* input, IntType offset, IntType prime) {
        zeek::ODesc desc(zeek::DESC_READABLE);
 
        auto length = 0;
@@ -943,14 +944,11 @@ uint64_t fnva(zeek::Val* input, uint64_t offset, uint64_t prime, FNV_HASH_TYPE t
                        length = desc.Len();
                }
 
-       uint64_t rval = offset;
+       IntType rval = offset;
        for ( auto i = 0; i < length; ++i )
        {
-               rval ^= (uint64_t) bytes[i];
+               rval ^= bytes[i];
                rval *= prime;
-
-               if (type == FNV1a32)
-                       rval = rval % (1ULL << 32);
        }
 
        return rval;
```

Use as follows:
```
@@ -967,10 +965,10 @@ uint64_t fnva(zeek::Val* input, uint64_t offset, uint64_t prime, FNV_HASH_TYPE t
 ## .. zeek:see:: hrw_weight
 function fnv1a32%(input: any%): count
        %{
-       uint64_t offset = 2166136261;
-       uint64_t prime = 16777619;
+       uint32_t offset = 2166136261;
+       uint32_t prime = 16777619;
 
-       uint64_t hash = fnva(input, offset, prime, FNV1a32);
+       auto hash = fnva<uint32_t>(input, offset, prime);
        return zeek::val_mgr->Count(hash);
```

Reply:
Made the change as you suggested!